### ISSUE TITLE
Place downloaded PHARs in a .tooly/ folder

### ISSUE THREAD

**Issue by hollodotme** [tommy-muehle/tooly-composer-script#4]
# This is a proposal

Instead of downloading the `.phar` files directly to `vendor/bin` and creating symlinks in the same place, I suggest to download the `.phar` files to a tooly-specific folder, e.g. `<ProjectDir>/.tooly/`, and create a symlink from there to `vendor/bin`. 

`.tooly/` should be the default name of the folder, which can be overwritten by a new config key in the extra section of composer.json like this:

```json
"extra": {
    "tooly-dir": "./.otherDir/",
    "tools": {
        "phpunit": {
            "url": "https://phar.phpunit.de/phpunit.phar",
            "only-dev": true
        }
    }
}
```

## Advantages

* Tooly would be able to prune obsolete `.phar` files, when removed from the extra/tools section in `composer.json`. This is currently not possible as there could be other files/symlinks under vendor/bin that were not created by tooly.
* There would be a tooly-specific place for any future use, e.g. storing meta data or a composer-like `.lock` file.

## Disadvantages

* One more folder in the project dir that needs to be added to `.gitignore`

## BC break

No.

**Comment by tommy-muehle** [comment]
@hollodotme 
The disadvantage part can bypass if you symlink the PHAR from "vendor/tm/tooly-composer-script/bin" to "vendor/bin" instead from a special folder.

But the main question for me is how tooly can detect that a tool is removed from extra-section and must be removed from vendor/bin but comes not from the require-dev part. 
Without a .lock file or something else. 

The only possible opinion that i see is to have only ".phar" binaries. So you can add for example phpunit in version 5.5 via tooly and phpunit in version 4.8 via require-dev part in your composer.json file.

But then you must handle to different binaries.
vendor/bin/phpunit -> 4.8
vendor/bin/phpunit.phar -> 5.5

You know what i mean?

**Comment by hollodotme** [comment]
About the pruning of tools, when configs were removed:

The tooly folder only contains the phar files that were configured in the extra section. So if you remove a phar from that section, you can prune all phars from the tooly folder that have no config in composer.json.

For example:

```json
{
  "extra": {
		"tools": {
			"codecept": {
				"url": "http://codeception.com/codecept.phar",
				"only-dev": true
			},
			"phpmetrics": {
				"url": "https://github.com/phpmetrics/PhpMetrics/raw/master/build/phpmetrics.phar",
				"only-dev": true
			}
    }
  }
}
```

Will result in 2 .phar files in the tooly folder:

```
- .tooly/
   |- codecept.phar
   `- phpmetrics.phar 
```

Now removing codeception from config:

```json
{
  "extra": {
		"tools": {
			"phpmetrics": {
				"url": "https://github.com/phpmetrics/PhpMetrics/raw/master/build/phpmetrics.phar",
				"only-dev": true
			}
    }
  }
}
```

Should result in 

```
- .tooly/
   `- phpmetrics.phar 
```

As only phars installed by tooly are placed in the tooly folder you can check which tools are obsolete by diffing the folder contents against the keys of the extra/tools section.

After this you can check for broken symlinks to tooly folder in vendor/bin and remove them too.

**Comment by hollodotme** [comment]
Regarding your suggestion to put the phars into `vendor/tm/tooly-composer-script/bin`:

I won't do that, because this folder is "hard to browse to" when you want to exclude it from indexing in your IDE (e.g. PhpStorm). I often need that because of code inspection issue "multiple class implementation". 

For example when you use a phar that uses Symfony/Console and your app also uses Symfony/Console then the code inspection complains about the multiple definition of Application or Command class, because the phar is indexed by PhpStorm.

If the tooly folder is placed in the main project directory it is easy to exclude it from indexing by with one right and one left click.

**Comment by tommy-muehle** [comment]
@hollodotme 
You cannot assume that anybody has an IDE with indexing and inspection functionality. 
So my suggestion makes the most sense. And all other libraries that i know do the same. Such as composer himself. (you can see it with composer in require-dev part) So sorry, you must make one click more. 😜 

In summary there are two changes for me right now:
* Check for obsolete PHAR's and remove it
* Handle symlink to composer's binary directory if PHAR are downloaded

I have an working prototype but there are missing tests at the time. 
But you can expect the release this week. (before vacation really begins 😁)

**Comment by hollodotme** [comment]
@tommy-muehle Noted the directory thing. Actually didn't know it is best practice already. ;)

Unfortunately turned out that I wouldn't have been able to provide a PR this week anyway... Project scheduled for Monday -.-

Thanks for your efforts!

**Comment by tommy-muehle** [comment]
@hollodotme 
Current [release branch](https://github.com/tommy-muehle/tooly-composer-script/tree/release/1.2.0) are stable and works. Only some tests are missing.
If you want test the result you can use this in your project:
```
"require": {
    "tm/tooly-composer-script": "dev-release/1.2.0#93b318e0dd31976a78844fff8915b4b77f1032d5"
}
```

**Comment by hollodotme** [comment]
@tommy-muehle Tested. Works like a charm. But... 😄 

a) The symlinks in `vendor/bin` now have the `.phar` extension, which is a BC break.

b) The symlinks have absolute source paths, which makes them marked as broken on the host system, when working on a guest OS. IMO the symlinks could have relative source paths, because the composer depths is fixed. 

For example: 

Current absolute symlink (broken on host, OK on guest):
```shell
cd vendor/bin
ls -l
[...] codecept.phar -> /vagrant/vendor/tm/tooly-composer-script/bin/codecept.phar
```

Relative symlink (works on both systems):
```
cd vendor/bin
ls -l
[...] codecept -> ../tm/tooly-composer-script/bin/codecept.phar
```

c) There should be a short info when a file/symlink is pruned. Like `Removing xyz`

**Comment by tommy-muehle** [comment]
@hollodotme 
Very big thanks for testing this release candidate and your feedback. Your point number two was fixed with 88c0d7e52d6f0591230df1d956f9627488e2e8a3. For point number 3 i created an separate issue #7. 

But point number 1 is discussable for me. Actually you get two PHAR's for each entry in the tools section. For example: "bin/tool.phar" and a copy of this under "bin/tool".
Is it really a BC break if only the PHAR remains?

Note that this gives you the option to add the tool also in the require-dev section of composer without overriding through the .phar in the same bin directory.

And you are no more ask for overriding the tool if any dependency comes with the same tool as requirement in any other version!

**Comment by hollodotme** [comment]
@tommy-muehle No big deal. 

Regarding Point 1: My team and I just got used to the execution of tools without the .phar extension , so for me it is a behaviour change - or my behaviour to be changed :) 
But anyway your arguments strike first. So, I'm fine if you'll leave it with the .phar extension.
